christfellowshipchurch/web-app
# [Chat] DMs Improvements & Chat Error Handling

# Related Todos
- [**🏕 🔥 [DMs] Show indicator for unread/new messages**](https://3.basecamp.com/3926363/buckets/17566548/todos/3005468554)
- [**🏕 🔥 [DMs] New DMs don't show in the list until refresh**](https://3.basecamp.com/3926363/buckets/17566548/todos/3076503485)
- [**🏕 🔥 [Auth] Logging in sometimes throws error**](https://3.basecamp.com/3926363/buckets/17566548/todos/3057764922)
- [**🏕 [Chat] Messages/input visual polish**](https://3.basecamp.com/3926363/buckets/17566548/todos/3076475123) (mostly, at least)
- [**🏕 [DMs] Update DMs banner design**](https://3.basecamp.com/3926363/buckets/17566548/todos/3071488250)

# Changes or Updates
- Major refactor of DMs dropdown to use Bootstrap styled dropdown and show unread indicators
- Listen to message received events to dynamically show new DM's and/or unread indicator
- Hide DM chats where last message received > 12hrs
- Better chat connection error handling, in lieu of more specific solution for intermittent logout/login as different user issue
- MessageInput styling refined
- Disable `/` and `:` and `@` commands for the time being

# Tradeoffs/Caveats

**Responding to new DMs**
This PR uses a heavy handed approach. When a new message is received in a DM channel, we re-fetch all of a users' DM channels. This can be refined later, but for now, is the simplest way to enforce consistency in the various components.

**Chat Error Handling**
Another heavy handed approach, the chat initialization logic is wrapped in a big `try/catch`. I've failed to isolate the _true_ cause of the intermittent Stream Chat errors, but they do seem to only happen when switching from one user to another.
Regardless, to prevent a hard crash of the entire page, a user-friendly error is shown and refreshing the page always fixes it.
This can also be refined later.

# Test Instructions
- Play with direct messages between two users (use an incognito window for second user)
- DMs where the last message received was > 12hrs ago should not appear.
  - For testing, you can change the time value in `src/stream-chat-client/chatUtils.js` on **line 92**
- If someone else initiates a DM with you, it should appear automatically/show the dropdown if it wasn't before
- Unread indicator should show on the **Direct Messages** dropdown as a whole, and on the individual DM channels
- Log out of one user, and in as another without refreshing the page. You _may_ get an error. If you do, verify the site is still usable, and that chat displays a user-friendly message.
- Verify the critical "happy path" for regular users — going from logged out => in state

# Screenshots

|Caption|Screenshot/GIF|
|---|---|
|**DMs Flow**|<img src="https://user-images.githubusercontent.com/1812955/95214934-67b49280-07be-11eb-8672-094ca6efab9c.gif" width="300" />|
|**DMs Dropdown Styling**|<img src="https://user-images.githubusercontent.com/1812955/95215097-9c284e80-07be-11eb-8e94-00e1463654d4.gif" width="300" />|
|**Updated MessageInput design**|<img src="https://user-images.githubusercontent.com/1812955/95215042-8ca90580-07be-11eb-9b72-b353e273a624.jpg" width="300" />|
|**Chat connection error state**|<img src="https://user-images.githubusercontent.com/1812955/95215479-03460300-07bf-11eb-8c12-b6cdf5ee0bb3.jpg" width="300" />|

# Tests
